pkg/server/server.go

```diff
@@ -73,9 +77,33 @@ func (c *ProxyClientConnection) send(pkt *client.Packet) error {
 }
 
 func NewPendingDialManager() *PendingDialManager {
-	return &PendingDialManager{
+	pm := &PendingDialManager{
 		pendingDial: make(map[int64]*ProxyClientConnection),
 	}
+
+	// spawn a goroutine to verify connection waiting is timeout or not.
+	go func() {
+		ticker := time.NewTicker(3 * time.Second)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				pm.mu.Lock()
+				for random, proxyClientConn := range pm.pendingDial {
+					// client connection pending over 15seconds
+					if time.Since(proxyClientConn.start) > ConnectionSetupTimeout {
+						delete(pm.pendingDial, random)
+						close(proxyClientConn.connected)
+						klog.V(2).InfoS("connection setup timeout", "random", random)
```

Comment:
metrics.Metrics.ObserveDialLatency(time.Since(pm.pendingDial.start)) is especially important in this use case. Would also be good if we had a DialError metric.